main/lsp/requests/code_action.cc

```diff
@@ -29,8 +29,229 @@ vector<unique_ptr<TextDocumentEdit>> getEdits(const LSPConfiguration &config, co
     }
     return documentEdits;
 }
+
+std::optional<const ast::MethodDef *> findMethodTree(const ast::ExpressionPtr &tree, const core::SymbolRef &method) {
+    if (auto seq = ast::cast_tree<ast::InsSeq>(tree)) {
+        for (auto &subtree : seq->stats) {
+            auto maybeMethod = findMethodTree(subtree, method);
+            if (maybeMethod.has_value()) {
+                return maybeMethod.value();
+            }
+        }
+    } else if (auto klass = ast::cast_tree<ast::ClassDef>(tree)) {
+        for (auto &subtree : klass->rhs) {
+            auto maybeMethod = findMethodTree(subtree, method);
+            if (maybeMethod.has_value()) {
+                return maybeMethod.value();
+            }
+        }
+    } else if (auto methodDef = ast::cast_tree<ast::MethodDef>(tree)) {
+        if (methodDef->symbol == method.asMethodRef()) {
+            return methodDef;
+        }
+    }
+    return std::nullopt;
+}
+
+unique_ptr<string> copyMethodSource(const core::GlobalState &gs, const core::LocOffsets &sigLoc,
+                                    const core::LocOffsets &methodLoc, const core::FileRef &fref) {
+    auto cutSource = [&](core::LocOffsets loc) {
+        return fref.data(gs).source().substr(loc.beginPos(), loc.endPos() - loc.beginPos());
+    };
+
+    auto sigSource = cutSource(sigLoc);
+    auto methodSource = cutSource(methodLoc);
+    return make_unique<string>(absl::StrCat(sigSource, "\n  ", methodSource));
+}
+
+std::optional<std::pair<core::LocOffsets, core::LocOffsets>> methodLocs(const core::GlobalState &gs,
+                                                                        const ast::ExpressionPtr &rootTree,
+                                                                        const core::SymbolRef &method,
+                                                                        const core::FileRef &fref) {
```

Comment:
```suggestion
                                                                        const core::SymbolRef method,
                                                                        const core::FileRef fref) {
```

These data structures are designed to be passed by copying.